kame/kame/shisad/cnd.8
.\"	$Id: cnd.8,v 1.1 2004/09/27 04:06:00 t-momose Exp $
.\"
.\" Copyright (C) 2004 WIDE Project.
.\" All rights reserved.
.\" 
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the project nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd June 2, 2004
.Dt CND 8
.Os KAME
.Sh NAME
.Nm cnd
.Nd Mobile IPv6 correspondent node daemon
.\"
.Sh SYNOPSIS
.Nm
.Op Fl d
.Op Fl i Ar interface
.\"
.Sh DESCRIPTION
.Nm
is a daemon to add a node Mobile IPv6 correspondent node capability.
.\"
.Sh HISTORY
The
.Nm
command first appeared in Shisa Mobile IPv6 protocol stack.
.Sh BUGS
All of parts.


CHANGELOG
			CHANGELOG for KAME kit
$KAME: CHANGELOG,v 1.2671 2004/10/05 07:42:00 suz Exp $

<200410>
Tue Oct 26 15:42:56 JST 2004    keiichi@iij.ad.jp
        * kame/sys/netinet6/icmp6.c: fixed a freebsd5 kernel crash due to a
        LOR caused by ICMPv6 redirect.

2004-10-21  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
        * kame/sys/netinet6/nd6_rtr.c (in6_ifadd): made sure that the
        "new" address does not happen to have been configured already.
        Otherwise, the same address can be doubly configured under some
        rare conditions.

2004-10-05  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/sys/netinet6/in6_rmx.c: fixed a freebsd5 kernel crash due to a
	  LOR caused by ICMPv6 redirect

2004-10-05  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* freebsd5/sys/dev/syscons/syscons.c: FreeBSD Security Advisory:
	  FreeBSD-SA-04:15.syscons 

2004-10-01  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/sys/netinet6/in6_rmx.c (in6_clsroute):
	* freebsd4/sys/netinet/in_rmx.c (in_clsroute):
	avoided duplicate free (and system crash) when
	net.inet[6].ip[6].rtexpire is set to 0.

<200409>
2004-09-29  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/sys/net,netinet,netinet6: removed a "Bad malloc flag" warning
	on freebsd5.

2004-09-07  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: provided a new command "dhcp6ctl", deprecating
	dhcp6sctl.  This new command controls the client as well as the
	server.  For the client, the "start" and "stop" (which can specify
	a particular interface) commands are available.

2004-09-04  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: added the ability of dhcp6c to work on multiple
	interfaces simultaneously.

<200408>
Thu Aug 12 12:54:25 JST 2004	keiichi@iij.ad.jp
	* kame/sys/netinet6/nd6.c,in6_src.c
	made it possible to specify the address of the other side of
	a p2p interface as nexthop when sending a packet with
	the NEXTHOP pktoption.

2004-08-11  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/ndp/ndp.c (ifinfo): supported the "disabled" ndp flag
	to view or modify the status of whether IPv6 operation is disabled
	on the interface. (see also the next log entry)

2004-08-11  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/sys/netinet6/nd6_nbr.c (nd6_dad_duplicated):
	* kame/sys/netinet6/ip6_input.c (ip6intr):
	* kame/sys/netinet6/nd6.c (nd6_output):
	disabled IPv6 operation if DAD fails for an address whose
	interface identifier is based on an EUI-64 ID which is supposed to
	be unique (a clarification in rfc2462bis).

2004-08-11  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/freebsd5/sbin/ifconfig/ifconfig.c: stopped unnecessary kernel
	module loading

2004-08-11  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/sys/netinet6/nd6_rtr.c (prelist_update): applied
	clarification on the prefix length check in rfc2462bis:
	- any prefix length is now accepted for on-link determination
	- determine the appropriate length based on the link type, rather
	  than on the prefix value (visible behavior should be same
	  though)

2004-08-10  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/kame/pmsft: protocol-independent Multicast-Source-Filter
	testing tool to test IGMPv3/MLDv2.

2004-08-01  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/freebsd5/usr.sbin/dhcp6: builds DHCPv6 in FreeBSD5

Sun Aug  1 04:07:45 2004  SUMIKAWA Munechika  <sumikawa@ebina.hitachi.co.jp>
	* freebsd4: Upgrade to 4.10-RELEASE

<200407>
2004-07-31  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/dhcp6c.c (main): made sure to dump the correct
	PID in the background mode.
	* kame/kame/dhcp6/dhcp6s.c (process_auth): corrected the timing to
	set the realm length.  Without this, dhcp6s could crash when
	processing authentication information.
	From: Kensaku Fujimoto

2004-07-30  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: relocated some configuration files.
	Automatically-generated DUID files are now in /var/db.  All the
	other configuration files should be put under /usr/local/v6/etc
	by default for consistency.
	Note: This change may affect the compatibility with old
	configurations.

2004-07-30  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/dhcp6s.c: allowed a Request/Renew/Release
	message to be unicasted when it is forwarded from a relay agent.
	From: Bilal Ahmed

2004-07-27  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/sys/net/if*.c: removed a "malloc() failure" warning message
 	  at start-up time on freebsd5.

Mon Jul 26 17:21:06 JST 2004	keiichi@iij.ad.jp
	* freebsd[45]/sys/netinet/tcp_syncache.c tcp_var.h:
	Initialize IPv6 flowinfo field in syncache/syncookie mechanism.

2004-07-26  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/kame/ping6/ping6.c: removed a one-second delay in the
	  ICMPv6 ECHO advertisement.
	* kame/sys/netinet6/nd6.c: fixed a freebsd5 kernel crash due to a
	  LOR in a TCPv6 session.

Wed Jul 21 13:14:08 JST 2004  itojun@iijlab.net
	* libinet6/name6.c: no need query ip6.int reverse tree any more.

2004-07-16  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/kame/kame/pim6sd/mld6v2_proto.c: fixed a wrong S-flag handling.
	* kame/sys/netinet6/icmp6.c: uses TCP-hostcache framework for pMTU
	  discovery in freebsd5, so that pMTU discovery on TCP works fine.

2004-07-14  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/sys/netinet6/in6.c: fixed a freebsd5 kernel crash by "ndp -P"
	command.

2004-07-12  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/sys/netinet6/in6.c: fixed a freebsd5 kernel crash due to LOR.
	* freebsd5/sys/netinet6/in6_pcb.c: fixed a bug that some TCP/UDP
	packets do not go outside on freebsd5.

2004-07-10  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* *bsd/usr.sbin/Makefile: stopped building KAME bgpd.  It has
	reached the end of life.

2004-07-09  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/kame/pim6sd/timer.c: fixed a bug that that MLDv1/v2 query 

kame/kame/shisad/cnd.c
/*	$Id: cnd.c,v 1.2 2004/09/27 08:50:43 t-momose Exp $	*/

/*
 * Copyright (C) 2004 WIDE Project.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the project nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <errno.h>
#include <syslog.h>
#include <poll.h>

#include <sys/param.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/sockio.h>

#include <net/if.h>
#if defined(__FreeBSD__) && __FreeBSD__ >= 3
#include <net/if_var.h>
#endif /* __FreeBSD__ >= 3 */
#include <netinet/in.h>
#include <netinet/icmp6.h>
#include <netinet/ip6mh.h>
#include <net/mipsock.h>
#include <netinet6/mip6.h>

#include "callout.h"
#include "stat.h"
#include "shisad.h"
#include "fsm.h"
#include "fdlist.h"
#include "command.h"

static void command_show_status(int, char *);
static void command_flush(int, char *);
static void terminate(int);

struct command_table command_table[] = {
	{"show", command_show_status, "Show status"},
	{"flush", command_flush, "Flush binding caches"},
};

/* Global Variables */
int mhsock, mipsock, icmp6sock;
int debug = 0, numerichost = 0;
struct mip6stat mip6stat;

static char *pid_file = CND_PIDFILE;

static void cn_lists_init(void);
void flush_bc(void);

void
cn_usage(path)
	char *path;
{
	char *cmd;

	cmd = strrchr(path, '/');
	if (!cmd)
		cmd = path;
	else
		cmd++;
	fprintf(stderr, "%s [-dn] [-i if]\n", cmd);
	exit(0);
} 


int
main(int argc, char **argv)
{
	int pfds;
	int pid;
	int ch = 0;
	char *arg_string;	/* XXX Bad var name; what is used for ? */
	FILE *pidfp;

#if 0
	/* XXX Is this check needed ? */
        if (argc < 2) {
		cn_usage(argv[0]);
		/* Not reach */
		return EINVAL;
	}
#endif

	/* get options */
	arg_string = NULL;
	while ((ch = getopt(argc, argv, "dni:")) != -1) {
		switch (ch) {
		case 'd':
			debug = 1;
			break;
		case 'n':
			numerichost = 1;
			break;
		case 'i':
			arg_string = optarg;
			break;
		default:
			fprintf(stderr, "unknown option\n");
			cn_usage();
			break;
		}
	}
	argc -= optind;
	argv += optind;

	/* open syslog infomation. */
	openlog("shisad(cnd)", 0, LOG_DAEMON);
	syslog(LOG_INFO, "-- Start CN daemon at %s -- \n", arg_string);

	/* open sockets */
	mhsock_open();
	mipsock_open();
	icmp6sock_open();

	/* start timer */
	callout_init();

	/* initialization */
	fdlist_init();
	if (command_init("cn> ", command_table, sizeof(command_table) / sizeof(struct command_table), 7777) < 0) {
		fprintf(stderr, "Unable to open user interface\n");
	}
	cn_lists_init();
	init_nonces();

	/* dump current PID */
	pid = getpid();
	if ((pidfp = fopen(pid_file, "w")) != NULL) {
		fprintf(pidfp, "%d\n", pid);
		fclose(pidfp);
	}

	/* register signal handlers. */
	signal(SIGTERM, terminate);
	signal(SIGINT, terminate);

	if (debug == 0)
		daemon(0, 0);

	new_fd_list(mhsock, POLLIN, mh_input_common);
        new_fd_list(mipsock, POLLIN, mipsock_input_common);
        new_fd_list(icmp6sock, POLLIN, icmp6_input_common);

	/* notify a kernel to behave as a correspondent node. */
	mipsock_nodetype_request(MIP6_NODETYPE_CORRESPONDENT_NODE, 1);

	while (1) {
		clear_revents();
	    
		if ((pfds = poll(fdl_fds, fdl_nfds, get_next_timeout())) < 0) {
			perror("poll");
			continue;
		}

		if (pfds != 0) {
			dispatch_fdfunctions(fdl_fds, fdl_nfds);
		}
		/* Timeout */
		callout_expire_check();
	}

	return (0);
}

static void
cn_lists_init()
{
	mip6_bc_init();

	return;
}


int
mipsock_input(struct mip_msghdr *miphdr)
{

	int err = 0;

	switch (miphdr->miph_type) {
	case MIPM_BE_HINT:
		mipsock_behint_input(miphdr);
		break;
	default:
		break;
	}

	return err;
}

static void
command_show_status(s, arg)
	int s;
	char *arg;
{
        char msg[1024];

	if (strcmp(arg, "bc") == 0) {
                sprintf(msg, "-- Binding Cache (Shisa) --\n");
                write(s, msg, strlen(msg));
		
                command_show_bc(s);

        } else if (strcmp(arg, "kbc") == 0) {
                sprintf(msg, "-- Binding Cache (kernel) --\n");
                write(s, msg, strlen(msg));

                command_show_kbc(s);

        } else if (strcmp(arg, "stat") == 0) {
                sprintf(msg, "-- Shisa Statistics --\n");
                write(s, msg, strlen(msg));

                command_show_stat(s);

        } else {
                sprintf(msg, "Available options are:\n");
                sprintf(msg + strlen(msg), "\tbc (Binding Cache in Shisa)\n\tkbc (Binding Cache in kernel)\n\tstat (Statistics)\n");
                write(s, msg, strlen(msg));
        }

	return;
}

static void
command_flush(s, arg)
	int s;
	char *arg;
{
	flush_bc();
}

static void
terminate(dummy)
	int dummy;
{
	mip6_flush_kernel_bc();
	unlink(pid_file);
	exit(1);
}

